instance/models/rabbitmq_server.py

```diff
@@ -0,0 +1,137 @@
+# -*- coding: utf-8 -*-
+#
+# OpenCraft -- tools to aid developing and hosting free software projects
+# Copyright (C) 2015-2016 OpenCraft <xavier@opencraft.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Instance app models - RabbitMQ Server model and manager
+"""
+
+# Imports #####################################################################
+
+import logging
+import random
+from urllib.parse import urlparse
+
+from django.conf import settings
+from django.core.exceptions import ImproperlyConfigured
+from django.db import models, transaction
+from django_extensions.db.models import TimeStampedModel
+
+from .utils import ValidateModelMixin
+
+
+# Logging #####################################################################
+
+logger = logging.getLogger(__name__)
+
+
+# Constants ###################################################################
+
+DEFAULT_INSTANCE_RABBITMQ_PORT = 5671
+
+
+# Models ######################################################################
+
+class RabbitMQServerManager(models.Manager):
+    """
+    Custom manager for the RabbitMQServer model.
+    """
+
+    def _create_default(self):
+        """
+        Create the default RabbitMQ server configured in the Django settings, if any.
+        """
+        if settings.DEFAULT_RABBITMQ_API_URL:
+            api_url = urlparse(settings.DEFAULT_RABBITMQ_API_URL)
+            amqps_url = urlparse(settings.DEFAULT_INSTANCE_RABBITMQ_URL)
+
+            defaults = {
+                'api_url': api_url.geturl().replace(':'.join([api_url.username, api_url.password]), ''),
```

Comment:
Ah, there's actually another problem here too, the aim was to pass the API URL through with the API credentials extracted (because these are already stored in separate fields), but I neglected to actually pass this through to the `get_or_create`. I've improved this so that this won't fail anymore from this line.